engine/src/main/java/org/hibernate/search/engine/metadata/impl/AnnotationMetadataProvider.java

```diff
@@ -1670,40 +1672,35 @@ private void checkForIndexedEmbedded(
 			return;
 		}
 
+		boolean elementCollection = isElementCollection( member );
+		if ( elementCollection ) {
+			// @IndexEmbedded used with @ElementCollection
+			logWarnings( member, indexedEmbeddedAnnotation );
+		}
+
 		parseContext.collectUnqualifiedCollectionRole( member.getName() );
 
 		int oldMaxLevel = parseContext.getMaxLevel();
-		int potentialLevel = depth( indexedEmbeddedAnnotation ) + parseContext.getLevel();
-		// This is really catching a possible int overflow. depth() can return Integer.MAX_VALUE, which then can
-		// overflow in case level > 0. Really this code should be rewritten (HF)
-		if ( potentialLevel < 0 ) {
-			potentialLevel = Integer.MAX_VALUE;
-		}
-		// HSEARCH-1442 recreating the behavior prior to PropertiesMetadata refactoring
+		int potentialLevel = elementCollection ? INFINITE_DEPTH : potentialLevel( parseContext, indexedEmbeddedAnnotation, member );
+		// HSEARCH-1442 recreating the behaviour prior to PropertiesMetadata refactoring
 		// not sure whether this is algorithmically correct though. @IndexedEmbedded processing should be refactored (HF)
 		if ( potentialLevel < oldMaxLevel ) {
 			parseContext.setMaxLevel( potentialLevel );
 		}
 		parseContext.incrementLevel();
 
-		XClass elementClass;
-		if ( void.class == indexedEmbeddedAnnotation.targetElement() ) {
-			elementClass = member.getElementClass();
-		}
-		else {
-			elementClass = reflectionManager.toXClass( indexedEmbeddedAnnotation.targetElement() );
-		}
+		XClass elementClass = elementCollection ? returnedType( member ) : elementClass( member, indexedEmbeddedAnnotation );
+		String localPrefix = elementCollection ? defaultPrefix( member ) : buildEmbeddedPrefix( prefix, indexedEmbeddedAnnotation, member );
+		boolean includeEmbeddedObjectId = elementCollection ? false : indexedEmbeddedAnnotation.includeEmbeddedObjectId();
```

Comment:
It might be just me so don't feel obliged to change it but I would have found it more readable with a global if (elementCollection) else, instead of 3 ternary operators.